runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/WindmillEndpoints.java

```diff
@@ -0,0 +1,192 @@
+      return new AutoValue_WindmillEndpoints_Endpoint.Builder();
+    }
+
+    public static Endpoint from(Windmill.WorkerMetadataResponse.Endpoint endpointProto) {
+      Endpoint.Builder endpointBuilder = Endpoint.builder();
+      if (endpointProto.hasDirectEndpoint() && !endpointProto.getDirectEndpoint().isEmpty()) {
+        parseDirectEndpoint(endpointProto.getDirectEndpoint())
+            .ifPresent(endpointBuilder::setDirectEndpoint);
+      }
+      if (endpointProto.hasWorkerToken() && !endpointProto.getWorkerToken().isEmpty()) {
+        endpointBuilder.setWorkerToken(endpointProto.getWorkerToken());
+      }
+
+      Endpoint endpoint = endpointBuilder.build();
+
+      if (!endpoint.directEndpoint().isPresent() && !endpoint.workerToken().isPresent()) {
+        throw new IllegalArgumentException(
+            String.format(
+                "direct_endpoint=[%s] not present or could not be parsed, and worker_token"
+                    + " not present. At least one of these fields is required.",
+                endpointProto.getDirectEndpoint()));
+      }
+
+      return endpoint;
+    }
+
+    private static Optional<WindmillServiceAddress> parseDirectEndpoint(String directEndpoint) {
+      Optional<WindmillServiceAddress> directEndpointIpV6Address =
+          tryParseDirectEndpointIntoIpV6Address(directEndpoint).map(WindmillServiceAddress::create);
+
+      return directEndpointIpV6Address.isPresent()
+          ? directEndpointIpV6Address
+          : tryParseEndpointIntoHostAndPort(directEndpoint).map(WindmillServiceAddress::create);
+    }
+
+    private static Optional<HostAndPort> tryParseEndpointIntoHostAndPort(String directEndpoint) {
+      try {
+        return Optional.of(HostAndPort.fromString(directEndpoint));
+      } catch (IllegalArgumentException e) {
+        LOG.warn("{} cannot be parsed into a gcpServiceAddress", directEndpoint);
+        return Optional.empty();
+      }
+    }
+
+    private static Optional<Inet6Address> tryParseDirectEndpointIntoIpV6Address(
+        String directEndpoint) {
+      InetAddress directEndpointAddress = null;
+      try {
+        directEndpointAddress = Inet6Address.getByName(directEndpoint);
+      } catch (UnknownHostException e) {
+        LOG.warn(
+            "Error occurred trying to parse direct_endpoint={} into IPv6 address. Exception={}",
+            directEndpoint,
+            e.toString());
+      }
+
+      // Inet6Address.getByAddress returns either an IPv4 or an IPv6 address depending on the format
+      // of the direct_endpoint string.
+      if (!(directEndpointAddress instanceof Inet6Address)) {
+        LOG.warn(
+            "{} is not an IPv6 address. Direct endpoints are expected to be in IPv6 format.",
+            directEndpoint);
+        return Optional.empty();
+      }
+
+      return Optional.ofNullable((Inet6Address) directEndpointAddress);
+    }
+
+    public abstract Optional<WindmillServiceAddress> directEndpoint();
```

Comment:
nit: move public accessors above all the private methods?
add some comments about how these shoudl be used or refer to the proto file and add a comment there?

Reply:
done. added comments and linked to proto and java files.